docs/architecture/ADR-001-ABCI++.md

```diff
@@ -0,0 +1,414 @@
+       panic(err)
+   }
+
+   eds, err := da.ExtendShares(squareSize, dataSquare)
+   if err != nil {
+       // ... log error
+       panic(err)
+   }
+
+   dah := da.NewDataAvailabilityHeader(eds)
+   data.Hash = dah.Hash() // <-- here we are setting the data hash before we pass the block data back to tendermint
+   data.OriginalSquareSize = squareSize
+
+   return abci.ResponsePrepareProposal{
+       BlockData: data,
+   }
+}
+```
+
+We estimate the square size by assuming that all the malleable transactions in the block have a valid commitment for whatever square size that we end up picking, and then quickly iterating through the block data to add up the expected lengths of each message/transaction. Please see [here](https://github.com/celestiaorg/celestia-app/blob/e18d8d2301a96702e1bf684735a3620eb059b12f/app/prepare_proposal.go#L47-L130) for more details.
+
+In order to efficiently fill the data square and ensure that each message included in the block is paid for, we progressively generate the data square using a few new types. More details can be found in [#637](https://github.com/celestiaorg/celestia-core/pull/637)
+
+```go
+// ContiguousShareWriter lazily merges transaction or other contiguous types in
+// the block data into shares that will eventually be included in a data square.
+// It also has methods to help progressively count how many shares the transactions
+// written take up.
+type ContiguousShareWriter struct {
+   shares       []NamespacedShare
+   pendingShare NamespacedShare
+   namespace    namespace.ID
+}
+...
+// MessageShareWriter lazily merges messages into shares that will eventually be
+// included in a data square. It also has methods to help progressively count
+// how many shares the messages written take up.
+type MessageShareWriter struct {
+   shares [][]NamespacedShare
+   count  int
+}
+```
+
+These types are combined in a new celestia-app type, `squareWriter`, which is responsible for atomically writing transactions and their corresponding messages to the data square and the returned block data.
+
+```go
+// squareWriter writes a data square using provided block data. It also ensures
+// that message and their corresponding txs get written to the square
+// atomically.
+type squareWriter struct {
+   txWriter  *coretypes.ContiguousShareWriter
+   msgWriter *coretypes.MessageShareWriter
+   ...
+}
+```
+
+```go
+// WriteSquare uses the provided block data to create a flattened data square.
+// Any MsgWirePayForMessages are malleated, and their corresponding
+// MsgPayForMessage and Message are written atomically. If there are
+// transactions that will node fit in the given square size, then they are
+// discarded. This is reflected in the returned block data. Note: pointers to
+// block data are only used to avoid dereferencing, not because we need the block
+// data to be mutable.
+func WriteSquare(txConf client.TxConfig, squareSize uint64, data *core.Data) ([][]byte, *core.Data, error) {
```

Comment:
nit: WriteSquare sounds like this actually writes the square into storage or sth.

Reply:
agree f7b6fcb7ba324d75f52a484585090c5844ca8a22